lib/Runtime/Library/JavascriptMap.cpp

```diff
@@ -6,462 +6,678 @@
+    case MapKind::ComplexVarMap:
     {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        MapDataNode * node = nullptr;
+        if (!this->u.complexVarMap->TryGetValue(key, &node))
+        {
+            return false;
+        }
+        *value = node->data.Value();
+        return true;
+    }
+    default:
+        Assume(UNREACHED);
+        return false;
+    }
+}
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
 
-        if (!JavascriptMap::Is(args[0]))
+bool
+JavascriptMap::Has(Var key)
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
+    switch (this->kind)
+    {
+    case MapKind::EmptyMap:
+        return false;
+
+    case MapKind::SimpleVarMap:
+    {
+        // First check if the key is in the map
+        if (this->u.simpleVarMap->ContainsKey(key))
         {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Map.prototype.keys"), _u("Map"));
+            return true;
+        }
+        // If the key isn't in the map, check if the canonical value is
+        Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(key);
+        // If the simple var is the same as the original key, we know it isn't in the map
+        if (!simpleVar || simpleVar == key)
+        {
+            return false;
         }
 
-        JavascriptMap* map = JavascriptMap::FromVar(args[0]);
-
-        return scriptContext->GetLibrary()->CreateMapIterator(map, JavascriptMapIteratorKind::Key);
+        return this->u.simpleVarMap->ContainsKey(simpleVar);
     }
+    case MapKind::ComplexVarMap:
+        return this->u.complexVarMap->ContainsKey(key);
 
-    Var JavascriptMap::EntryValues(RecyclableObject* function, CallInfo callInfo, ...)
-    {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+    default:
+        Assume(UNREACHED);
+        return false;
+    }
+}
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
+void
+JavascriptMap::PromoteToComplexVarMap()
+{
+    AssertOrFailFast(this->kind == MapKind::SimpleVarMap);
 
-        if (!JavascriptMap::Is(args[0]))
-        {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Map.prototype.values"), _u("Map"));
-        }
+    uint newMapSize = this->u.simpleVarMap->Count() + 1;
+    ComplexVarDataMap* newMap = RecyclerNew(this->GetRecycler(), ComplexVarDataMap, this->GetRecycler(), newMapSize);
 
-        JavascriptMap* map = JavascriptMap::FromVar(args[0]);
-        return scriptContext->GetLibrary()->CreateMapIterator(map, JavascriptMapIteratorKind::Value);
+    JavascriptMap::MapDataList::Iterator iter = this->list.GetIterator();
+    // TODO: we can use a more efficient Iterator, since we know there will be no side effects
+    while (iter.Next())
```

Comment:
>Next [](start = 16, length = 4)

There is comment "Nodes can be deleted while iterating so validate current" in MapOrSetDataList.h Next() does that have any effect?

Reply:
in general case, this iterator can be exposed to users, so elements might get deleted during iteration, but here that can't happen

---
In reply to: [174325329](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174325329) [](ancestors = 174325329)